File: auth/laravel/app/Services/VibeNestAuth.php
<?php

namespace App\Services;

use Firebase\JWT\JWK;
use Firebase\JWT\JWT;
use Illuminate\Support\Facades\Http;
use RuntimeException;

final class VibeNestAuth
{
    public function __construct(
        private readonly string $issuer,
        private readonly string $clientId,
        private readonly ?string $clientSecret,
        private readonly string $redirectUri,
    ) {
        if (parse_url($issuer, PHP_URL_SCHEME) !== 'https' || parse_url($redirectUri, PHP_URL_SCHEME) !== 'https') {
            throw new RuntimeException('OIDC issuer and callback must use HTTPS.');
        }
    }

    public static function fromEnvironment(): self
    {
        return new self(
            rtrim(self::required('VIBENEST_AUTH_ISSUER'), '/').'/',
            self::required('VIBENEST_AUTH_CLIENT_ID'),
            env('VIBENEST_AUTH_CLIENT_SECRET') ?: null,
            self::required('VIBENEST_AUTH_REDIRECT_URI'),
        );
    }

    public function begin(array $metadata): array
    {
        $this->assertMetadata($metadata);
        $verifier = self::base64Url(random_bytes(64));
        $pending = [
            'verifier' => $verifier,
            'state' => self::base64Url(random_bytes(32)),
            'nonce' => self::base64Url(random_bytes(32)),
            'expires_at' => time() + 300,
        ];
        $query = http_build_query([
            'client_id' => $this->clientId,
            'redirect_uri' => $this->redirectUri,
            'response_type' => 'code',
            'scope' => 'openid profile email',
            'code_challenge' => self::base64Url(hash('sha256', $verifier, true)),
            'code_challenge_method' => 'S256',
            'state' => $pending['state'],
            'nonce' => $pending['nonce'],
        ], '', '&', PHP_QUERY_RFC3986);
        return ['url' => $metadata['authorization_endpoint'].'?'.$query, 'pending' => $pending];
    }

    public function complete(array $query, array $pending, array $metadata): array
    {
        $this->assertMetadata($metadata);
        if (($pending['expires_at'] ?? 0) < time()
            || !isset($query['code'], $query['state'], $pending['state'], $pending['verifier'], $pending['nonce'])
            || !hash_equals($pending['state'], $query['state'])) {
            throw new RuntimeException('Invalid or expired OIDC callback.');
        }
        $form = [
            'grant_type' => 'authorization_code',
            'code' => $query['code'],
            'client_id' => $this->clientId,
            'redirect_uri' => $this->redirectUri,
            'code_verifier' => $pending['verifier'],
        ];
        if ($this->clientSecret !== null) {
            $form['client_secret'] = $this->clientSecret;
        }
        $token = Http::asForm()->timeout(10)->post($metadata['token_endpoint'], $form)->throw()->json();
        $jwks = Http::timeout(10)->get($metadata['jwks_uri'])->throw()->json();
        JWT::$leeway = 30;
        $claims = (array) JWT::decode($token['id_token'], JWK::parseKeySet($jwks));
        $audience = is_array($claims['aud'] ?? null) ? $claims['aud'] : [$claims['aud'] ?? null];
        if (($claims['iss'] ?? null) !== $this->issuer
            || !in_array($this->clientId, $audience, true)
            || !isset($claims['sub'], $claims['exp'], $claims['iat'], $claims['nonce'])
            || !hash_equals($pending['nonce'], (string) $claims['nonce'])) {
            throw new RuntimeException('OIDC ID token validation failed.');
        }
        $profile = [];
        if (isset($token['access_token'], $metadata['userinfo_endpoint'])) {
            $profile = Http::withToken($token['access_token'])->timeout(10)->get($metadata['userinfo_endpoint'])->throw()->json();
            if (($profile['sub'] ?? null) !== $claims['sub']) {
                throw new RuntimeException('UserInfo subject mismatch.');
            }
        }
        return ['subject' => $claims['sub'], 'email' => $profile['email'] ?? $claims['email'] ?? null];
    }

    public function discovery(): array
    {
        $metadata = Http::timeout(10)->get($this->issuer.'.well-known/openid-configuration')->throw()->json();
        $this->assertMetadata($metadata);
        return $metadata;
    }

    public function applicationOrigin(): string
    {
        return parse_url($this->redirectUri, PHP_URL_SCHEME).'://'.parse_url($this->redirectUri, PHP_URL_HOST)
            .(parse_url($this->redirectUri, PHP_URL_PORT) ? ':'.parse_url($this->redirectUri, PHP_URL_PORT) : '');
    }

    private function assertMetadata(array $metadata): void
    {
        if (($metadata['issuer'] ?? null) !== $this->issuer) {
            throw new RuntimeException('OIDC discovery issuer mismatch.');
        }
    }

    private static function base64Url(string $value): string
    {
        return rtrim(strtr(base64_encode($value), '+/', '-_'), '=');
    }

    private static function required(string $name): string
    {
        $value = trim((string) env($name, ''));
        if ($value === '') {
            throw new RuntimeException("Missing {$name}.");
        }
        return $value;
    }
}

File: auth/rails/app/services/vibe_nest_auth.rb
require "base64"
require "digest"
require "json"
require "jwt"
require "net/http"
require "openssl"
require "securerandom"
require "uri"

class VibeNestAuth
  attr_reader :issuer, :client_id, :client_secret, :redirect_uri

  def initialize(issuer:, client_id:, client_secret:, redirect_uri:, http: Net::HTTP)
    @issuer = issuer.end_with?("/") ? issuer : issuer + "/"
    @client_id = client_id
    @client_secret = client_secret.presence
    @redirect_uri = redirect_uri
    @http = http
    raise ArgumentError, "OIDC issuer and callback must use HTTPS" unless URI(issuer).scheme == "https" && URI(redirect_uri).scheme == "https"
  end

  def self.from_environment
    new(
      issuer: ENV.fetch("VIBENEST_AUTH_ISSUER"),
      client_id: ENV.fetch("VIBENEST_AUTH_CLIENT_ID"),
      client_secret: ENV["VIBENEST_AUTH_CLIENT_SECRET"],
      redirect_uri: ENV.fetch("VIBENEST_AUTH_REDIRECT_URI")
    )
  end

  def begin_flow(metadata)
    validate_metadata!(metadata)
    verifier = base64url(SecureRandom.random_bytes(64))
    pending = { "verifier" => verifier, "state" => base64url(SecureRandom.random_bytes(32)), "nonce" => base64url(SecureRandom.random_bytes(32)), "expires_at" => Time.now.to_i + 300 }
    query = URI.encode_www_form(
      client_id: client_id, redirect_uri: redirect_uri, response_type: "code",
      scope: "openid profile email", code_challenge: base64url(Digest::SHA256.digest(verifier)),
      code_challenge_method: "S256", state: pending["state"], nonce: pending["nonce"]
    )
    { url: metadata.fetch("authorization_endpoint") + "?" + query, pending: pending }
  end

  def complete(query, pending, metadata)
    validate_metadata!(metadata)
    unless pending && pending["expires_at"].to_i >= Time.now.to_i && query["code"].present? && secure_equal(pending["state"], query["state"])
      raise ArgumentError, "Invalid or expired OIDC callback"
    end
    form = { grant_type: "authorization_code", code: query.fetch("code"), client_id: client_id, redirect_uri: redirect_uri, code_verifier: pending.fetch("verifier") }
    form[:client_secret] = client_secret if client_secret
    token = post_form(metadata.fetch("token_endpoint"), form)
    jwks = get_json(metadata.fetch("jwks_uri"))
    claims, = JWT.decode(token.fetch("id_token"), nil, true,
      algorithms: %w[RS256 ES256], jwks: jwks, iss: issuer, verify_iss: true,
      aud: client_id, verify_aud: true, required_claims: %w[exp iat sub nonce])
    raise ArgumentError, "OIDC nonce mismatch" unless secure_equal(pending["nonce"], claims["nonce"])
    profile = token["access_token"] && metadata["userinfo_endpoint"] ? get_json(metadata["userinfo_endpoint"], bearer: token["access_token"]) : {}
    raise ArgumentError, "UserInfo subject mismatch" if profile["sub"] && profile["sub"] != claims["sub"]
    { "subject" => claims.fetch("sub"), "email" => profile["email"] || claims["email"] }
  end

  def discovery
    get_json(issuer + ".well-known/openid-configuration").tap { |metadata| validate_metadata!(metadata) }
  end

  def application_origin
    uri = URI(redirect_uri)
    "#{uri.scheme}://#{uri.host}#{uri.port == 443 ? '' : ":#{uri.port}"}"
  end

  private

  def validate_metadata!(metadata)
    raise ArgumentError, "OIDC discovery issuer mismatch" unless metadata["issuer"] == issuer
  end

  def get_json(url, bearer: nil)
    uri = URI(url); request = Net::HTTP::Get.new(uri); request["Authorization"] = "Bearer #{bearer}" if bearer
    response = @http.start(uri.host, uri.port, use_ssl: true, open_timeout: 5, read_timeout: 10) { |connection| connection.request(request) }
    raise "OIDC HTTP request failed" unless response.is_a?(Net::HTTPSuccess)
    JSON.parse(response.body)
  end

  def post_form(url, form)
    uri = URI(url); request = Net::HTTP::Post.new(uri); request.set_form_data(form)
    response = @http.start(uri.host, uri.port, use_ssl: true, open_timeout: 5, read_timeout: 10) { |connection| connection.request(request) }
    raise "OIDC token request failed" unless response.is_a?(Net::HTTPSuccess)
    JSON.parse(response.body)
  end

  def base64url(value) = Base64.urlsafe_encode64(value, padding: false)
  def secure_equal(left, right) = left.is_a?(String) && right.is_a?(String) && left.bytesize == right.bytesize && ActiveSupport::SecurityUtils.secure_compare(left, right)
end
